src/components/tl/dpu/tl_dpu_coll.c

```diff
@@ -0,0 +1,213 @@
+{
+    ucc_tl_dpu_task_t *task         = ucc_derived_of(coll_task, ucc_tl_dpu_task_t);
+    ucc_tl_dpu_team_t *team         = task->team;
+    ucc_tl_dpu_context_t *ctx       = UCC_TL_DPU_TEAM_CTX(team);
+
+    ucp_request_param_t req_param;
+    tl_dpu_ucp_request_t *recv_req;
+    volatile uint32_t *check_flag = team->ctrl_seg;
+
+    if (team->coll_id != (*check_flag + 1)) {
+        return UCC_INPROGRESS;
+    }
+    req_param.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                             UCP_OP_ATTR_FIELD_DATATYPE;
+    req_param.datatype     = ucp_dt_make_contig(1);
+    req_param.cb.recv      = recv_handler_nbx;
+
+    recv_req = ucp_get_nbx(ctx->ucp_ep, task->args.dst.info.buffer,
+                           task->args.dst.info.count * ucc_dt_size(task->args.dst.info.datatype),
+                           team->rem_data_out, team->rem_data_out_key,
+                           &req_param);
+    if (tl_dpu_ucp_req_check(recv_req) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+    do {
+        ucp_worker_progress(ctx->ucp_worker);
+    } while((tl_dpu_ucp_req_test(&recv_req, ctx->ucp_worker) != UCC_OK));
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_dpu_allreduce_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_dpu_task_t *task     = ucs_derived_of(coll_task, ucc_tl_dpu_task_t);
+    ucc_tl_dpu_team_t *team     = task->team;
+    ucc_tl_dpu_context_t *ctx   = UCC_TL_DPU_TEAM_CTX(team);
+    
+    ucp_request_param_t req_param;
+    tl_dpu_ucp_request_t *put_req[2];
+
+    fprintf(stdout, "Collective post");
+    req_param.op_attr_mask  = UCP_OP_ATTR_FIELD_CALLBACK |
+                                UCP_OP_ATTR_FIELD_DATATYPE;
+    req_param.datatype      = ucp_dt_make_contig(1);
+    req_param.cb.send       = send_handler_nbx;
+    // req_param.mask          = 0/* XXX */;
+    // req_param.mem_type      = task->args.src.info.mem_type;
+    // req_param.memory_type   = ucc_memtype_to_ucs[mtype]; /* XXX */
+
+    /* IN PLACE CHECK special in place func in ucc_coll_utils */
+    put_req[0] = ucp_put_nbx(ctx->ucp_ep,
+                                task->args.src.info.buffer,
+                                task->args.src.info.count * ucc_dt_size(task->args.src.info.datatype),
+                                team->rem_data_in, team->rem_data_in_key, &req_param);
+    if (tl_dpu_ucp_req_check(put_req[0]) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+    ucp_worker_fence(ctx->ucp_worker);
+
+    put_req[1] = ucp_put_nbx(ctx->ucp_ep, &task->sync, sizeof(task->sync),
+                              team->rem_ctrl_seg, team->rem_ctrl_seg_key,
+                              &req_param);
+    if (tl_dpu_ucp_req_check(put_req[1]) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    do {
+        ucp_worker_progress(ctx->ucp_worker);
+    } while((tl_dpu_ucp_req_test(&(put_req[0]), ctx->ucp_worker) != UCC_OK) ||
+            (tl_dpu_ucp_req_test(&(put_req[1]), ctx->ucp_worker) != UCC_OK));
```

Comment:
This can block if RMA is not TRUE on-sided. Need to add proper non-blocking progress there in the future.

Reply:
indeed, I made the team-create NB, will do so here too